cluster-autoscaler/core/scale_up_test.go

```diff
@@ -970,3 +1067,112 @@ func TestCheckScaleUpDeltaWithinLimits(t *testing.T) {
 		}
 	}
 }
+
+func TestScaleUpLateBinding(t *testing.T) {
+	// A pod using an unbound PVC with late binding can be scheduled, so
+	// a cluster will get scaled up.
+	config := &scaleTestConfig{
+		nodes: []nodeConfig{
+			{"n1", 100, 100, 0, true, "ng1"},
+			{"n2", 1000, 1000, 0, true, "ng2"},
+		},
+		extraPods: []podConfig{
+			{"p-new", 500, 0, 0, "", false, []string{"pvc1"}},
+		},
+		pvcs: []pvcConfig{
+			{"pvc1", "1Gi", "lvm-storage-class"},
+		},
+		scs: []scConfig{
+			{"lvm-storage-class", "lvm.example.com"},
+		},
+		options:                 defaultOptions,
+		expansionOptionToChoose: groupSizeChange{groupName: "ng2", sizeChange: 1},
+	}
+	expectedResults := &scaleTestResults{
+		finalOption: groupSizeChange{groupName: "ng2", sizeChange: 1},
+		scaleUpStatus: scaleUpStatusInfo{
+			podsTriggeredScaleUp: []string{"p-new"},
+		},
+	}
+
+	simpleScaleUpTest(t, config, expectedResults)
+}
+
+func TestScaleUpNoStorage(t *testing.T) {
+	defer utilfeaturetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.CSIStorageCapacity, true)()
+
+	// When storage capacity tracking is enabled, a pod using an
+	// unbound PVC with late binding can be scheduled only if the
+	// node candidates are known to have capacity available, which
+	// isn't the case here.
+	config := &scaleTestConfig{
+		nodes: []nodeConfig{
+			{"n1", 100, 100, 0, true, "ng1"},
+			{"n2", 1000, 1000, 0, true, "ng2"},
+		},
+		extraPods: []podConfig{
+			{"p-new", 500, 0, 0, "", false, []string{"pvc1"}},
+		},
+		pvcs: []pvcConfig{
+			{"pvc1", "1Gi", "lvm-storage-class"},
+		},
+		scs: []scConfig{
+			{"lvm-storage-class", "lvm.example.com"},
+		},
+		csi: []csiDriverConfig{
+			{"lvm.example.com", true},
+		},
+		options:                 defaultOptions,
+		expansionOptionToChoose: groupSizeChange{groupName: "ng2", sizeChange: 1},
+	}
+	expectedResults := &scaleTestResults{
+		finalOption: groupSizeChange{},
+		scaleUpStatus: scaleUpStatusInfo{
+			podsRemainUnschedulable: []string{"p-new"},
+		},
+	}
+
+	simpleNoScaleUpTest(t, config, expectedResults)
+}
+
+func TestScaleUpEnoughStorage(t *testing.T) {
+	defer utilfeaturetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.CSIStorageCapacity, true)()
+
+	// When storage capacity tracking is enabled, a pod using an
+	// unbound PVC with late binding can be scheduled only if the
+	// node candidates are known to have capacity available.
+	// This can be configured for node candidates by giving them
+	// special labels (TODO: how?) and then manually creating
```

Comment:
This is the key conceptual question that I have: when cluster autoscaler creates node candidates, is it configurable which node labels those have?

This is not just relevant for the CSIStorageCapacity object below, but also for topology-aware pod scheduling. Suppose a PV has been created with topology constraints as provided by a CSI driver. How can the scheduler know that the node candidate is in the same topology segment as the existing volume? That's based on node labels, so those must be get set somehow.

Reply:
CA will copy an existing node to use as a node template. Any topology labels should be copied as well (CA assumes a single NodeGroup is made of nodes that are identical for scheduling purposes, ie. they're in one zone). 
For scale from 0 it's up to cloudprovider TemplateNodeInfo to provide a template node that has all the relevant labels.

Reply:
Looks like I need to read up on node groups. Do you perhaps have a link to documentation about how to configure those?

> CA will copy an existing node to use as a node template.

How does it pick that node?

> Any topology labels should be copied as well (CA assumes a single NodeGroup is made of nodes that are identical for scheduling purposes, ie. they're in one zone).

Hmm, I can imagine that it might be more complex. For example, the same zone (= data center) might be able to provide machines with different hardware features (for example, optional CPU instructions like AVX) and those features might be exposed as node labels.

When considering node-local storage, each node is in its own topology zone. Cloning the labels of an existing node would be wrong for the labels set for a CSI driver on that node.

Reply:
> Looks like I need to read up on node groups. Do you perhaps have a link to documentation about how to configure those?

A NodeGroup is an internal cluster autoscaler concept for a resizable set of identical nodes. How it maps to underlying infrastructure will be different for each cloud provider (ex. on AWS a NodeGroup is an ASG, on GCP it's a ManagedInstanceGroup) and the way to set it up will also differ by cloudprovider. The instructions for most providers can be found in cluster-autoscaler/cloudprovider/<provider_name>/README.md.

> How does it pick that node?

Randomly. The assumption is that a NodeGroup is a set of identical nodes, so it doesn't matter which one CA uses as a template for representing future nodes.

> Hmm, I can imagine that it might be more complex. For example, the same zone (= data center) might be able to provide machines with different hardware features (for example, optional CPU instructions like AVX) and those features might be exposed as node labels.

Two nodes using different hardware should be part of different NodeGroups. Cluster Autoscaler can autoscale cluster with multiple NodeGroups.

A very high level description of how CA works is:
* A cluster is made of different NodeGroups, which represent some sort of resizable collection of VMs in underlying cloud. All CA ever does to actuate its decisions is resize NodeGroup. The expectation is that underlying cloud will in that case provide more identical VMs and those VMs will self-register in the cluster.
* A cluster can be made of multiple NodeGroups. Some or all of those may be autoscaled.
* If multiple NodeGroups are autoscaled CA will run a separate binpacking simulation for each of those.
  * To perform a simulation it will create an in-memory Node object that represents a new node that can be added to the cluster. It will than use embedded scheduler code to in-memory "schedule" pending pods on this node. CA will keep adding more in-memory nodes (copies of original template node) until all the pending pods can be "scheduled". The outcome of this simulation is how many nodes would need to be added to a particular NodeGroup to make all the pending pods schedulable.
  * Once the simulation is done CA will reset the simulation by deleting all in-memory nodes and erasing any "scheduling" decisions it took. It will than run simulation for next NodeGroup. This process will be repeated for each NodeGroup in the cluster.
  * A configurable heuristic ("expander") will look at results of simulations and pick one of them. CA will increase target size of selected NodeGroup (ie. ask underlying cloud for more VMs).
 
> When considering node-local storage, each node is in its own topology zone. Cloning the labels of an existing node would be wrong for the labels set for a CSI driver on that node.

I think I'm missing some context here. Is the CSIStorageCapacity object for a node or for some class of identical nodes?

Reply:
Note how the algorithm described above doesn't work with features like local PVs (since scheduler code takes PVs from informer and not snapshot, there is no way for CA to inject an in-memory PV object that represents the local volume on in-memory node).

Reply:
> I think I'm missing some context here. Is the CSIStorageCapacity object for a node or for some class of identical nodes?

That depends on the storage system. For network-attached storage, it may be for multiple nodes in a certain zone because all those nodes share the same underlying storage. For local storage, each node has its own storage and once the node is up, will have CSIStorageCapacity objects that represent storage on that node.

If all nodes in a NodeGroup are alike and have the same amount of storage capacity, then one CSIStorageCapacity object can describe those nodes. We just have to take care that objects for nodes that don't exist yet and nodes that exits are not in conflict with each other. That can be achieved by labeling the node candidates differently - not something that CA currently supports, but it doesn't sound hard to add.

This is then using the existing APIs of the volume binder (informers for CSIStorageCapacity, PVC, Node, CSIDriver, etc.). If we change those APIs, then other solutions are also possible.

Reply:
> Note how the algorithm described above doesn't work with features like local PVs (since scheduler code takes PVs from informer and not snapshot, there is no way for CA to inject an in-memory PV object that represents the local volume on in-memory node).

This is where we need different solutions for different kind of storage systems. For local storage managed by a CSI driver with dynamic provisioning, there won't be any such local PVs and there won't be a need to inject fake PVs. Instead, we need to inform the scheduler about capacity on the non-existent nodes. I was focusing on that approach.

I don't know which approach will be more common in clusters where CA will be used. :shrug: